osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -325,38 +322,160 @@ private void checkFirmwareHistory(final String deviceId,
+        // moduleType
+        Map<String, FirmwareVersionWithInstallationDate> currentlyInstalledFirmwareVersionsPerType = new HashMap<String, FirmwareVersionWithInstallationDate>();
+
+        for (DeviceFirmwareFile firmwareFile : deviceFirmwareFiles) {
+
+            Map<FirmwareModule, String> fwms = firmwareFile.getFirmwareFile().getModuleVersions();
+            Date installationDate = firmwareFile.getInstallationDate();
+
+            for (Map.Entry<FirmwareModule, String> entry : fwms.entrySet()) {
+                String version = entry.getValue();
+                FirmwareModule fwm = entry.getKey();
+                if (currentlyInstalledFirmwareVersionsPerType.containsKey(fwm.getDescription())) {
+                    // check if this installation of this same kind of module is
+                    // of a later date
+                    if (currentlyInstalledFirmwareVersionsPerType.get(fwm.getDescription())
+                            .get_installationDate()
+                            .before(installationDate)) {
+                        currentlyInstalledFirmwareVersionsPerType.replace(fwm.getDescription(),
+                                new FirmwareVersionWithInstallationDate(installationDate, new FirmwareVersion(
+                                        FirmwareModuleType.forDescription(fwm.getDescription()), version)));
+                    }
+                } else {
+                    // no other module of this type found yet so just add it
+                    currentlyInstalledFirmwareVersionsPerType.put(fwm.getDescription(),
+                            new FirmwareVersionWithInstallationDate(installationDate, new FirmwareVersion(
+                                    FirmwareModuleType.forDescription(fwm.getDescription()), version)));
+                }
+            }
+        }
+
+        final List<FirmwareVersion> latestfirmwareVersionsOfEachModuleTypeInHistory = currentlyInstalledFirmwareVersionsPerType
+                .values()
+                .stream()
+                .map(e -> new FirmwareVersion(e.get_firmwareVersion().getFirmwareModuleType(),
+                        e.get_firmwareVersion().getVersion()))
+                .collect(Collectors.toList());
+
+        // remove the latest history (module)versions from the firmwareVersions
+        // parameter
+        firmwareVersionsToCheck.removeAll(latestfirmwareVersionsOfEachModuleTypeInHistory);
+
+        return firmwareVersionsToCheck;
+    }
+
+    // Helper class to keep track of InstallationDate and FirmwareVersion
+    private class FirmwareVersionWithInstallationDate {
+        private Date _installationDate;
+        private FirmwareVersion _firmwareVersion;
+
+        public Date get_installationDate() {
+            return _installationDate;
+        }
+
+        public void set_installationDate(Date _installationDate) {
+            this._installationDate = _installationDate;
+        }
+
+        public FirmwareVersion get_firmwareVersion() {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Rename this method name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS00100)